### ISSUE TITLE
[IDEA]  Always create unique titles for tiddlers being imported

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#6560]
**Problem**
Tiddlers don't always get unique titles when they're imported. This risks overwriting existing tiddlers and it burdens the user to make up a title. One example of this is when you paste multiple *images* from the clipboard: the import mechanism proposes the title `image.png` even when that tiddler already exists.

**Request**
Importing - regardless of whether it is done from a paste or drag and drop event - should always create unique tiddler titles. It is fine if the user can also opt to modify the title but the systemically proposed default should be unique.

**Comment by saqimtiaz** [comment]
> Importing - regardless of whether it is done from a paste or drag and drop event - should always create unique tiddler titles.

This would need to be a setting users can toggle in the import UI, as it wont always be a desirable behaviour.

**Comment by twMat** [comment]
@saqimtiaz 

Thanks. I modified the very last sentence in the OP. Does it cover your concern now?

**Comment by saqimtiaz** [comment]
@twMat not quite. The new behaviour should be something that can be switched on and off during import, which is not the same as allowing the user to modify the title. Imagine importing a JSON file with a bunch of tiddlers, more likely than not you do not want those tiddlers to be renamed and given unique titles.

I would rephrase the last paragraph as:
>Importing - regardless of whether it is done from a paste or drag and drop event - should offer an option to create unique tiddler titles. This setting should be controllable from the import UI.

**Comment by twMat** [comment]
> Imagine importing a JSON file with a bunch of tiddlers, more likely than not you do not want those tiddlers to be renamed and given unique titles.

Right...

>[...] should offer an option to create unique tiddler titles. This setting should be controllable from the import UI.

But isn't this how it is currently? As you know, my [original](https://github.com/Jermolene/TiddlyWiki5/issues/6559) frustration gave the example how pasting clipboard text creates  `Untitled`, `Untitled 1`, `Untitled 2` ... but pasting clipboard images repeatedly creates  `image.png`. There *is* already an import mechanism UI option to rename it, so my point is that the mechanism should treat images and strings consistently and give them all unique titles by *default*.

It seems you want to go for an opt-in for unique titles whereas I'm proposing an opt-in for non-unique titles. Maybe we can identify some characteristic that indicates which is best as default; a unique title or non unique title. For example, the content from the clipboard that I have in mind are not tiddlers (e.g no .tid file suffix) whereas shadowtids in a JSON obviously are tiddlers and they already have titles.

**Comment by saqimtiaz** [comment]
> But isn't this how it is currently?

I am thinking of a toggle that makes all titles being imported unique, rather than needing to individually rename tiddlers queued for import. An easy to use toggle should also diminish the importance of which default setting for that toggle that a particular user prefers.

**Comment by twMat** [comment]
>I am thinking of a toggle that makes all titles [...]

I agree that'd be a nice improvement - but perhaps not a huge one since the user still, for each import session, has to keep track of if he wants unique titles or not. What do you think of the idea to:

>identify some characteristic that indicates which is best as default; a unique title or non unique title.

...after all, that seems to be how it *is* done currently when pasting from clipboard: texts get unique titles but images don't. Could it be done sytematically? Maybe the user could then specify for your toggle that "default when pasting images should be unique titles", etc.

**Comment by pmario** [comment]
> the import mechanism proposes the title image.png even when that tiddler already exists.

The next bundler-plugin will have a function for this, which can be activated by the user

**Comment by joshuafontany** [comment]
There are overlapping concerns here, so from a Test Coverage perspective, which is more important:

A) Titles in the current $:/Import should not clash with any other titles in the current $:/Import.

B) Titles in the current $:/Import should not clash with any other titles in the Wiki.

Is it A, or B, or both A&B you want? Or some toggkeabke combination of them?

**Comment by CodaCodr** [comment]
B) Priority

A) nice to have - prefer last occurrence wins.

**Comment by rmunn** [comment]
For one of my use cases, I have a set of data in another format and I'm converting it to a .json file that I can import into TW. As I work out the conversion process, I will import the same tiddlers repeatedly from different .json files output from different builds of my conversion script. In my use case, I *want* the imported tiddlers to overwrite the tiddlers that already exist. And since there are hundreds of them, I don't want to have to click one checkbox per tiddler title to opt out of the "rename imported tiddlers to have unique titles" feature; I want to be able to click one button to choose the import mode, either "rename to have unique titles" mode or "incoming tiddlers overwrite existing ones" mode.

I'd be fine with "rename to unique titles" as the default, as long as it's a one-button click (or one checkbox) to switch to the "overwrite existing tiddlers with the same title" mode instead.

**Comment by joshuafontany** [comment]
Ok. And to clarify, B) should only apply to raw-data imports. Anything recognized as a tiddler (structured with a title field & recognizable by one of the tiddler parsers) should not be renamed, otherwise you brak the plugin upgrade process, and the standalone wiki upgrade process.

**Comment by rmunn** [comment]
> Anything recognized as a tiddler (structured with a title field & recognizable by one of the tiddler parsers) should not be renamed (by default), otherwise you break the plugin upgrade process, and the standalone wiki upgrade process.

That would be enough to satisfy my use case needs, because I'm producing a .json file in tiddler format (the .json file is a list of objects that have `title`, `text`, `tags`, and other properties).

**Comment by twMat** [comment]
How about this:

For *system* tiddlers, the global toggle is by default set to *overwrite / do not rename*, and
for *non-system* tiddlers the global toggle is by default set to *enumerate / rename*
For a mix of system and non-system tiddlers, the global toggle is by default set to e.g *overwrite*

**Comment by pmario** [comment]
See: https://talk.tiddlywiki.org/t/changing-image-png-for-importing-images/2698/27?u=pmario

I did just update all my plugins to TW 5.2.2 and the bundler-plugin got a new import feature. ... 

@twMat 
It may be possible to use the same, or a similar mechanism for `untitled` tiddlers and it should be possible to convert text content, that looks like a tiddler into a tiddler, if the user wants that

**Comment by twMat** [comment]
@pmario - as noted in the thread you refer to, I posted about some issues in your gh for bundler.

>It may be possible to use the same, or a similar mechanism for untitled tiddlers and it should be possible to convert text content, that looks like a tiddler into a tiddler, if the user wants that

That sounds very good. Is it something you're experimenting with?

**Comment by pmario** [comment]
@twMat ... I think we can close this one. The core import now allows us to edit every tiddler title on import, if needed.
